# E-Commerce Customer & Sales Analysis — Turkey

**Dataset:** E-commerce transaction data (Jan 2023 – Mar 2024)  
**Tools:** MySQL, Python, Pandas, SQLAlchemy, Jupyter Notebook  

## Objective
Analyze customer retention, buying behavior, delivery performance and 
revenue trends to understand what drives customer loyalty and business growth.

### Connection Cell

In [3]:
from sqlalchemy import create_engine
import pandas as pd

engine = create_engine("mysql+mysqlconnector://root:aqsam1022@localhost/ecoomerce")

def run_query(query):
    return pd.read_sql(query, engine)

## Section: Customer Retention Analysis

### Q1 — Category wise retention

In [4]:
run_query("""
SELECT product_category,
    COUNT(CASE WHEN Is_Returning_Customer = 'True' THEN 1 END) as returning_customers,
    COUNT(CASE WHEN Is_Returning_Customer = 'False' THEN 1 END) as new_customers,
    ROUND(COUNT(CASE WHEN Is_Returning_Customer = 'True' THEN 1 END) * 100.0 / COUNT(*), 2) as retention_rate
FROM ecommerce
GROUP BY product_category
ORDER BY retention_rate DESC
""")

,product_category,returning_customers,new_customers,retention_rate
0,Electronics,386,238,61.86
1,Food,380,239,61.39
2,Home & Garden,370,251,59.58
3,Toys,363,247,59.51
4,Fashion,370,252,59.49
5,Books,365,251,59.25
6,Sports,395,272,59.22
7,Beauty,361,260,58.13


**Finding:** Retention rates are fairly close across categories, ranging from 
58% to 62%. Electronics (61.86%) and Food (61.39%) have the highest retention, 
suggesting customers buying these categories come back most often. Beauty 
has the lowest retention at 58.13%, though the gap between best and worst 
is small — only about 3.7 percentage points — meaning category alone is not 
a strong driver of retention in this dataset.

### Q2 — Gender wise retention

In [7]:
run_query("""
SELECT product_category,
    ROUND(COUNT(CASE WHEN Is_Returning_Customer = 'True' AND gender = 'Male' THEN 1 END) * 100.0 /
          COUNT(CASE WHEN gender = 'Male' THEN 1 END), 2) as male_retention_rate,
    ROUND(COUNT(CASE WHEN Is_Returning_Customer = 'True' AND gender = 'Female' THEN 1 END) * 100.0 /
          COUNT(CASE WHEN gender = 'Female' THEN 1 END), 2) as female_retention_rate,
    ROUND(COUNT(CASE WHEN Is_Returning_Customer = 'True' AND gender = 'Other' THEN 1 END) * 100.0 /
          COUNT(CASE WHEN gender = 'Other' THEN 1 END), 2) as other_retention_rate
FROM ecommerce
GROUP BY product_category
ORDER BY male_retention_rate DESC
""")

,product_category,male_retention_rate,female_retention_rate,other_retention_rate
0,Food,63.21,59.52,58.33
1,Home & Garden,61.48,58.13,50.00
2,Electronics,59.61,64.14,61.54
3,Books,59.50,59.64,46.67
4,Sports,59.41,58.82,71.43
5,Fashion,58.68,60.67,40.00
6,Toys,57.73,60.90,71.43
7,Beauty,56.95,58.81,75.00


**Finding:** Retention patterns differ noticeably by gender within categories. 
Men show the highest retention in Food (63.21%) and Home & Garden (61.48%), 
while women show highest retention in Electronics (64.14%). The "Other" 
gender group shows more extreme swings, likely due to a much smaller sample 
size in that group, which makes those percentages less statistically reliable 
(e.g. 75% in Beauty vs 40% in Fashion). Overall, male and female retention 
rates stay close to each other across most categories, generally within 
3-5 percentage points.

### Q3 — Age wise retention

In [8]:
run_query("""
SELECT
    CASE
        WHEN Age BETWEEN 18 AND 25 THEN '18-25'
        WHEN Age BETWEEN 26 AND 35 THEN '26-35'
        WHEN Age BETWEEN 36 AND 45 THEN '36-45'
        WHEN Age BETWEEN 46 AND 55 THEN '46-55'
        ELSE '55+'
    END as age_group,
    COUNT(*) as total_customers,
    COUNT(CASE WHEN Is_Returning_Customer = 'True' THEN 1 END) as returning_customers,
    ROUND(COUNT(CASE WHEN Is_Returning_Customer = 'True' THEN 1 END) * 100.0 / COUNT(*), 2) as retention_rate
FROM ecommerce
GROUP BY age_group
ORDER BY retention_rate DESC
""")

,age_group,total_customers,returning_customers,retention_rate
0,46-55,697,431,61.84
1,36-45,1470,888,60.41
2,18-25,1125,675,60.00
3,55+,194,115,59.28
4,26-35,1514,881,58.19


**Finding:** The 46-55 age group shows the highest retention rate at 61.84%, 
followed closely by 36-45 (60.41%) and 18-25 (60.00%). The 26-35 age group, 
despite having the largest customer base (1514 customers), has the lowest 
retention at 58.19%. This suggests middle-aged and older customers are 
slightly more loyal than younger adult customers, though the overall spread 
across all age groups is narrow  about 3.6 percentage points between 
highest and lowest.

## Section: Buying Patterns

### Q4 — Most bought category by gender

In [9]:
run_query("""
SELECT product_category, gender,
    SUM(Quantity) as total_quantity,
    ROUND(SUM(Total_Amount), 2) as total_revenue,
    RANK() OVER (PARTITION BY product_category ORDER BY SUM(Total_Amount) DESC) as revenue_rank
FROM ecommerce
GROUP BY product_category, gender
ORDER BY product_category, revenue_rank
""")

,product_category,gender,total_quantity,total_revenue,revenue_rank
0,Beauty,Male,696.0,77899.95,1
1,Beauty,Female,702.0,76660.43,2
2,Beauty,Other,26.0,2024.36,3
3,Books,Male,645.0,36455.31,1
4,Books,Female,604.0,34398.12,2
5,Books,Other,42.0,1891.09,3
6,Electronics,Male,683.0,1162335.36,1
7,Electronics,Female,672.0,1134289.67,2
8,Electronics,Other,24.0,32181.78,3
9,Fashion,Male,714.0,187906.28,1


**Finding:** Men generate more revenue than women in most categories, 
particularly Beauty, Books, Electronics, Fashion, and Food despite Beauty 
and Electronics not being stereotypically "male" categories. Women dominate 
revenue in Home & Garden (476K vs 423K), Sports (412K vs 336K), and Toys 
(113K vs 108K). The "Other" gender group contributes a small but consistent 
share across all categories, generally under 3% of total revenue. Electronics 
is by far the highest revenue category overall for both genders (over 1.1M 
combined), making it the most commercially important category in the dataset.

### Q5 — Most bought category by age group

In [10]:
run_query("""
WITH age_group_category AS (
    SELECT 
        CASE
            WHEN Age BETWEEN 18 AND 25 THEN '18-25'
            WHEN Age BETWEEN 26 AND 35 THEN '26-35'
            WHEN Age BETWEEN 36 AND 45 THEN '36-45'
            WHEN Age BETWEEN 46 AND 55 THEN '46-55'
            ELSE '55+'
        END as age_group,
        product_category,
        SUM(Quantity) as Total_Quantity
    FROM ecommerce
    GROUP BY product_category, age_group
)
SELECT product_category, age_group, Total_Quantity,
    ROUND(Total_Quantity * 100.0 / SUM(Total_Quantity) OVER (PARTITION BY product_category), 2) as total_percentage
FROM age_group_category
ORDER BY product_category, total_percentage DESC
""")

,product_category,age_group,Total_Quantity,total_percentage
0,Beauty,26-35,446.0,31.32
1,Beauty,36-45,379.0,26.62
2,Beauty,18-25,325.0,22.82
3,Beauty,46-55,208.0,14.61
4,Beauty,55+,66.0,4.63
5,Books,26-35,442.0,34.24
6,Books,36-45,367.0,28.43
7,Books,18-25,296.0,22.93
8,Books,46-55,142.0,11.00
9,Books,55+,44.0,3.41


**Finding:** The 26-35 and 36-45 age groups dominate purchasing volume across 
almost every category, together accounting for roughly 55-65% of total 
quantity sold in each category. Sports shows the strongest skew toward the 
36-45 group (33.31%), while Books and Toys are slightly more popular with 
26-35 and 36-45 nearly tied. The 55+ age group consistently makes up the 
smallest share in every category, never exceeding 6% of total purchases. 
This suggests the core customer base across all product categories is 
working-age adults between 26 and 45, rather than younger or older shoppers.

## Section: Delivery Analysis

### Q6 — Average delivery time per category and city

In [11]:
run_query("""
SELECT product_category,
    ROUND(AVG(Delivery_Time_Days), 2) as avg_delivery_days
FROM ecommerce
GROUP BY product_category
ORDER BY avg_delivery_days DESC
""")

,product_category,avg_delivery_days
0,Sports,6.71
1,Food,6.63
2,Home & Garden,6.53
3,Books,6.47
4,Fashion,6.45
5,Toys,6.42
6,Electronics,6.39
7,Beauty,6.37


**Finding:** Delivery times are remarkably consistent across all categories, 
ranging narrowly from 6.37 days (Beauty) to 6.71 days (Sports) only about 
8 hours difference between fastest and slowest. This suggests delivery 
performance is driven more by logistics infrastructure or city than by 
the type of product being shipped.

### Q7 — Delivery time vs customer rating

In [12]:
run_query("""
SELECT 
    CASE 
        WHEN Delivery_Time_Days BETWEEN 1 AND 5 THEN '1-5 days'
        WHEN Delivery_Time_Days BETWEEN 6 AND 10 THEN '6-10 days'
        WHEN Delivery_Time_Days BETWEEN 11 AND 15 THEN '11-15 days'
        WHEN Delivery_Time_Days BETWEEN 16 AND 20 THEN '16-20 days'
        ELSE '20+ days'
    END as delivery_bucket,
    ROUND(AVG(Customer_Rating), 2) as avg_rating,
    COUNT(*) as total_orders
FROM ecommerce
GROUP BY delivery_bucket
ORDER BY delivery_bucket
""")

,delivery_bucket,avg_rating,total_orders
0,1-5 days,3.92,2280
1,11-15 days,3.87,556
2,16-20 days,3.90,82
3,20+ days,3.92,13
4,6-10 days,3.89,2069


**Finding:** Customer rating stays nearly flat regardless of delivery time, 
hovering between 3.87 and 3.92 across all buckets even orders taking 
20+ days still average 3.92, identical to orders delivered in 1-5 days. 
This contradicts the common assumption that slower delivery hurts 
satisfaction. In this dataset, delivery speed does not appear to meaningfully 
influence customer rating; other factors like product quality likely play 
a bigger role in how customers rate their experience.

## Section: Revenue Analysis

### Q8 — Month by month total revenue

In [13]:
run_query("""
SELECT YEAR(Date) as Year, MONTH(Date) as Month,
    ROUND(SUM(Total_Amount), 1) as monthly_revenue
FROM ecommerce
GROUP BY Year, Month
ORDER BY Year, Month
""")

,Year,Month,monthly_revenue
0,2023,1,286694.3
1,2023,2,338784.9
2,2023,3,321716.0
3,2023,4,257332.6
4,2023,5,351902.0
5,2023,6,324357.0
6,2023,7,346618.0
7,2023,8,391505.9
8,2023,9,262602.9
9,2023,10,352286.6


**Finding:** Revenue fluctuates month to month without an obvious strong 
seasonal pattern, ranging from a low of 257K (April 2023) to a high of 
396K (January 2024). August 2023 (391K) and December 2023 (384K) were 
also strong months, possibly tied to back-to-school and holiday shopping. 
January 2024 stands out as the highest revenue month in the entire dataset.

### Q9 — Best performing category each month

In [14]:
run_query("""
WITH revenue AS (
    SELECT product_category, YEAR(Date) as Year, MONTH(Date) as Month,
        ROUND(SUM(Total_Amount), 0) as Total_Revenue
    FROM ecommerce
    GROUP BY Year, Month, product_category
),
category AS (
    SELECT product_category, Year, Month, Total_Revenue,
        DENSE_RANK() OVER (PARTITION BY Year, Month ORDER BY Total_Revenue DESC) as Top
    FROM revenue
)
SELECT product_category, Year, Month, Total_Revenue, Top
FROM category
WHERE Top = 1
ORDER BY Year, Month
""")

,product_category,Year,Month,Total_Revenue,Top
0,Electronics,2023,1,141705.0,1
1,Electronics,2023,2,173686.0,1
2,Electronics,2023,3,162755.0,1
3,Electronics,2023,4,101217.0,1
4,Electronics,2023,5,164093.0,1
5,Electronics,2023,6,155524.0,1
6,Electronics,2023,7,174172.0,1
7,Electronics,2023,8,197490.0,1
8,Electronics,2023,9,106775.0,1
9,Electronics,2023,10,163347.0,1


**Finding:** Electronics was the top revenue-generating category in every 
single month from January 2023 through March 2024 without exception. This 
shows Electronics is not just a strong category but the dominant one no 
seasonal shift caused another category (like Toys in December or Fashion 
in summer) to overtake it at any point in the 15-month period.

### Q10 — Year over year comparison

In [15]:
run_query("""
WITH data AS (
    SELECT MONTH(Date) as Month, YEAR(Date) as Year, Total_Amount
    FROM ecommerce
),
monthly_data AS (
    SELECT Month,
        ROUND(SUM(CASE WHEN Year = 2023 THEN Total_Amount END), 0) as previous,
        ROUND(SUM(CASE WHEN Year = 2024 THEN Total_Amount END), 0) as current
    FROM data
    GROUP BY Month
)
SELECT Month, previous, current,
    CASE 
        WHEN current IS NULL THEN 'No 2024 data yet'
        WHEN previous > current THEN 'decline'
        WHEN previous < current THEN 'increase'
        ELSE 'no difference'
    END as progress
FROM monthly_data
ORDER BY Month
""")

,Month,previous,current,progress
0,1,286694.0,396875.0,increase
1,2,338785.0,342913.0,increase
2,3,321716.0,274633.0,decline
3,4,257333.0,NaN,No 2024 data yet
4,5,351902.0,NaN,No 2024 data yet
5,6,324357.0,NaN,No 2024 data yet
6,7,346618.0,NaN,No 2024 data yet
7,8,391506.0,NaN,No 2024 data yet
8,9,262603.0,NaN,No 2024 data yet
9,10,352287.0,NaN,No 2024 data yet


**Finding:** For the 3 months where both years can be compared, January and 
February 2024 both showed revenue increases over 2023 (January: 286K → 397K, 
+38%; February: 339K → 343K, +1%), while March 2024 declined compared to 
March 2023 (322K → 275K, -15%). With only 3 comparable months available, 
this is too small a sample to confidently conclude the business is growing 
overall January's strong jump is promising, but March's decline tempers 
that optimism. A fuller year of 2024 data would be needed for a reliable 
trend conclusion.

## Section: Returning Customer Behavior

### Q11 — Spending, rating and payment method differences

In [16]:
run_query("""
SELECT Is_Returning_Customer,
    ROUND(AVG(Total_Amount), 2) as avg_spending,
    ROUND(AVG(Customer_Rating), 2) as avg_rating,
    ROUND(AVG(Unit_Price), 2) as avg_unit_price
FROM ecommerce
GROUP BY Is_Returning_Customer
""")

,Is_Returning_Customer,avg_spending,avg_rating,avg_unit_price
0,True,985.96,3.90,465.83
1,False,978.87,3.91,440.96


In [17]:
run_query("""
SELECT Is_Returning_Customer, Payment_Method,
    COUNT(*) as total_orders,
    ROUND(COUNT(*) * 100.0 / SUM(COUNT(*)) OVER (PARTITION BY Is_Returning_Customer), 2) as percentage
FROM ecommerce
GROUP BY Is_Returning_Customer, Payment_Method
ORDER BY Is_Returning_Customer, percentage DESC
""")

,Is_Returning_Customer,Payment_Method,total_orders,percentage
0,False,Credit Card,825,41.04
1,False,Debit Card,488,24.28
2,False,Digital Wallet,385,19.15
3,False,Bank Transfer,210,10.45
4,False,Cash on Delivery,102,5.07
5,True,Credit Card,1187,39.70
6,True,Debit Card,777,25.99
7,True,Digital Wallet,580,19.40
8,True,Bank Transfer,300,10.03
9,True,Cash on Delivery,146,4.88


**Finding:** Returning customers spend slightly more on average (985.96 TRY 
vs 978.87 TRY) and buy marginally higher priced items (465.83 vs 440.96 TRY 
unit price), but the difference is small under 1% in spending and about 
6% in unit price. Customer ratings are essentially identical (3.90 vs 3.91). 
Payment method preference is also nearly the same across both groups
Credit Card is the top choice for both (41.04% new, 39.70% returning), 
followed by Debit Card and Digital Wallet in the same order. Overall, 
returning customers do not show meaningfully different purchasing behavior 
from new customers in this dataset.

### Q12 — Session behavior of returning vs new customers

In [18]:
run_query("""
SELECT Is_Returning_Customer,
    ROUND(AVG(Session_Duration_Minutes), 2) as avg_session_duration,
    ROUND(AVG(Pages_Viewed), 2) as avg_pages_viewed
FROM ecommerce
GROUP BY Is_Returning_Customer
""")

,Is_Returning_Customer,avg_session_duration,avg_pages_viewed
0,True,14.46,8.99
1,False,14.75,8.97


In [19]:
run_query("""
SELECT session_bucket, Is_Returning_Customer, total_orders,
    ROUND(total_orders * 100.0 / SUM(total_orders) OVER (PARTITION BY Is_Returning_Customer), 2) as pct_within_group
FROM (
    SELECT 
        CASE 
            WHEN Session_Duration_Minutes BETWEEN 1 AND 15 THEN '1-15 min'
            WHEN Session_Duration_Minutes BETWEEN 16 AND 30 THEN '16-30 min'
            WHEN Session_Duration_Minutes BETWEEN 31 AND 60 THEN '31-60 min'
            ELSE '60+ min'
        END as session_bucket,
        Is_Returning_Customer,
        COUNT(*) as total_orders
    FROM ecommerce
    GROUP BY session_bucket, Is_Returning_Customer
) as t
ORDER BY session_bucket, Is_Returning_Customer
""")

,session_bucket,Is_Returning_Customer,total_orders,pct_within_group
0,1-15 min,False,1214,60.40
1,1-15 min,True,1875,62.71
2,16-30 min,False,685,34.08
3,16-30 min,True,953,31.87
4,31-60 min,False,111,5.52
5,31-60 min,True,160,5.35
6,60+ min,True,2,0.07


**Finding:** Average session duration and pages viewed are nearly identical 
between returning (14.46 min, 8.99 pages) and new customers (14.75 min, 
8.97 pages). Looking at the distribution, returning customers are slightly 
more concentrated in short sessions under 15 minutes (62.71% vs 60.40%), 
but the difference is marginal about 2 percentage points. This does not 
strongly support the hypothesis that returning customers browse less because 
they already know what they want; both groups behave very similarly in 
terms of browsing time and depth.

### Q13 — Discount impact on retention

In [20]:
run_query("""
SELECT Is_Returning_Customer,
    COUNT(*) as total_orders,
    COUNT(CASE WHEN Discount_Amount > 0 THEN 1 END) as orders_with_discount,
    ROUND(COUNT(CASE WHEN Discount_Amount > 0 THEN 1 END) * 100.0 / COUNT(*), 2) as pct_orders_with_discount
FROM ecommerce
GROUP BY Is_Returning_Customer
""")

,Is_Returning_Customer,total_orders,orders_with_discount,pct_orders_with_discount
0,True,2990,906,30.30
1,False,2010,620,30.85


In [21]:
run_query("""
SELECT Is_Returning_Customer,
    ROUND(AVG(Discount_Amount), 2) as avg_discount_given
FROM ecommerce
GROUP BY Is_Returning_Customer
""")

,Is_Returning_Customer,avg_discount_given
0,True,25.46
1,False,23.95


**Finding:** Discount usage rate is nearly identical between returning 
(30.30%) and new customers (30.85%), and average discount amount is also 
similar (25.46 TRY vs 23.95 TRY). This suggests discounts are not a 
meaningful driver of customer retention in this dataset returning 
customers are not receiving discounts more frequently or generously than 
new customers. Whatever brings customers back is likely related to factors 
outside this dataset, such as product satisfaction, brand trust, or service 
quality, rather than discount incentives.

## Key Findings Summary

1. **Retention is broadly consistent** across categories, genders, and age 
groups — no single factor strongly predicts whether a customer returns
2. **Electronics is the dominant category**, topping revenue every single 
month across the entire 15-month period
3. **Delivery time has no real impact on customer rating** — even 20+ day 
deliveries rate the same as 1-5 day deliveries
4. **26-45 age group is the core customer base** across nearly every category
5. **Men generate more revenue** in most categories except Home & Garden, 
Sports, and Toys, where women lead
6. **Discounts do not drive retention** — usage rates are nearly identical 
between new and returning customers
7. **Returning customers behave almost identically to new customers** in 
spending, session duration, and payment preferences — loyalty in this 
dataset appears unrelated to browsing behavior or financial incentives

---
*Analysis by Aqsam | Tools: MySQL, Python, Power BI*